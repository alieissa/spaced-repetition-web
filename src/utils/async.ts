/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
