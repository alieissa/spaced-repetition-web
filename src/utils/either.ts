/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
