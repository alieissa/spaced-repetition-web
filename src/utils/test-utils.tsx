/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
