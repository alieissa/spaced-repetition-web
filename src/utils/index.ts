/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
