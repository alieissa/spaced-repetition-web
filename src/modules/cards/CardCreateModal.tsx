/** @format */

import { useEffect } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { Message } from 'semantic-ui-react'
import {
  SPButton,
  SPModal,
  SPModalActions,
  SPModalContent,
  SPModalHeader,
} from 'src/components'
import { async } from 'src/utils'
import CardForm from './CardForm'
import { useCardCreate, useCardForm } from './cards.hooks'
import { NCards } from './cards.types'

/**
 * This component contains the create card modal and the button that is used to toggle
 * the modal.
 */
export default function CardCreateModal() {
  const params = useParams()
  const navigate = useNavigate()
  const [createCardStatus, createCard] = useCardCreate(params.deckId!)

  const initCard = NCards.Initial({})

  const {
    form,
    getQuestionError,
    getAnswerError,
    handleAddAnswer,
    handleDeleteAnswer,
    handleChangeAnswer,
    handleChangeQuestion,
  } = useCardForm(initCard, () => {
    createCard(form.values)
  })

  const handleCloseModal = () => navigate(-1)

  useEffect(() => {
    if (createCardStatus.type === 'Success') {
      handleCloseModal()
    }
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [createCardStatus.type])

  return (
    <SPModal
      data-testid="card-create-modal"
      open={true}
      onClose={handleCloseModal}
    >
      <SPModalHeader>Create Card</SPModalHeader>
      <SPModalContent className="flex-column align-center justify-center">
        {async.match(createCardStatus)({
          Untriggered: () => null,
          Success: () => null,
          Loading: () => null,
          Failure: () => (
            <Message
              data-testid="card-create-error"
              negative
              className="w-inherit"
            >
              <Message.Header>Error: Failed to create card</Message.Header>
            </Message>
          ),
        })}

        <CardForm
          {...form.values}
          areAnswersVisible={true}
          getQuestionError={getQuestionError}
          getAnswerError={getAnswerError}
          onAddAnswer={handleAddAnswer}
          onChangeAnswer={handleChangeAnswer}
          onDeleteAnswer={handleDeleteAnswer}
          onChangeQuestion={handleChangeQuestion}
        />
      </SPModalContent>
      <SPModalActions>
        <SPButton onClick={handleCloseModal}>Cancel</SPButton>
        <SPButton
          data-testid="card-create-save-btn"
          color="green"
          onClick={form.submitForm}
        >
          Save
        </SPButton>
      </SPModalActions>
    </SPModal>
  )
}
