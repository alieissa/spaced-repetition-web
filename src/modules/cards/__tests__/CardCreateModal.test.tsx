/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/lib/node'
import { flushPromises, renderModalWithProviders } from 'src/utils/test-utils'
import CardCreateModal from '../CardCreateModal'

const deckId = faker.string.uuid()
const cardCreateUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}/cards`

const server = setupServer()
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

const mountComponent = () =>
  renderModalWithProviders(<CardCreateModal />, {
    initialEntries: [`/decks/${deckId}/cards/new`],
    path: 'decks/:deckId/cards/new',
  })

const fillInForm = async () => {
  const user = userEvent.setup()

  const questionInput = await screen.findByTestId('question-content')
  await act(() => user.type(questionInput as Element, faker.lorem.lines(1)))

  const answerInput = await screen.findByTestId('answer-content-0')
  await act(() => user.type(answerInput as Element, faker.lorem.lines(1)))
}

describe('CardCreateModal', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = mountComponent()

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it('should display error when card creation fails', async () => {
      // Assemble
      server.use(
        rest.post(cardCreateUrl, (__, res, ctx) => res(ctx.status(422))),
      )
      mountComponent()
      await fillInForm()

      // Act
      const cardCreateSaveBtn = screen.getByTestId('card-create-save-btn')
      await act(() => cardCreateSaveBtn.click())
      await act(flushPromises)

      // Assert
      const cardCreateError = await screen.findByTestId('card-create-error')
      expect(cardCreateError).toBeInTheDocument()
    })
  })
})
