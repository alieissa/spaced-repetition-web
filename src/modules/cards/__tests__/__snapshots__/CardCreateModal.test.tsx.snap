// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCreateModal view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui inverted top aligned page modals dimmer transition visible active"
    style="display: flex;"
  >
    <div
      class="ui modal transition visible active"
      data-testid="card-create-modal"
      style="box-shadow: none; border: 1px solid; border-radius: 0;"
    >
      <div
        class="header"
      >
        Create Card
      </div>
      <div
        class="flex-column align-center justify-center content"
      >
        <form
          class="ui form w-full bordered"
        >
          <div
            class="flex-row-reverse"
          >
            <i
              aria-hidden="true"
              class="arrow up icon"
              data-testid="view-answers-btn"
            />
          </div>
          <div
            class="ui horizontal list flex"
            role="list"
            style="display: flex;"
          >
            <div
              class="item flex-1"
              role="listitem"
              style="padding: 0px;"
            >
              <div
                class="field w-full"
              >
                <div
                  class="ui input"
                  style="opacity: 1;"
                >
                  <input
                    data-testid="question-content"
                    placeholder="Enter question here"
                    style="border-radius: unset;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="item flex-1"
              role="listitem"
              style="padding: 0px;"
            >
              <div
                class="ui list"
                role="list"
                style="padding: 0px;"
              >
                <div
                  class="item flex"
                  role="listitem"
                  style="display: flex; padding: 0px;"
                >
                  <div
                    class="field w-full"
                  >
                    <div
                      class="ui input"
                      style="opacity: 1;"
                    >
                      <input
                        data-testid="answer-content-0"
                        placeholder="Enter answer here"
                        style="border-radius: unset;"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <button
                    class="ui small icon disabled button"
                    disabled=""
                    style="padding-left: 0px; padding-right: 0px; background: none;"
                    tabindex="-1"
                  >
                    <i
                      aria-hidden="true"
                      class="x icon"
                    />
                  </button>
                </div>
                <div
                  class="item"
                  role="listitem"
                  style="text-align: right; padding: 0px;"
                >
                  <button
                    class="ui small icon button"
                    style="padding-left: 0px; padding-right: 0px; background: none;"
                  >
                    <i
                      aria-hidden="true"
                      class="plus icon"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
      <div
        class="actions"
      >
        <button
          class="ui button"
          style="border-radius: 0;"
        >
          Cancel
        </button>
        <button
          class="ui green button"
          data-testid="card-create-save-btn"
          style="border-radius: 0;"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
